NETBEANS-59 - Document split actions

Chris2011 opened the pull request:
Cleaned up code as discussed here: #1 

Everything fine now?

mcdonnell-john commented:
Did this need a second PR?  Could it not have been done in the first one?

Chris2011 commented:
Maybe not, but I reverted everything, cloned the repo new, pulled from the original master, created a new branch, copied over the files manually and change the line endings.

emilianbold commented:
Of course, it's much easier to read since there are no more newline changes.

What this patch seems to do is add an @ActionReference so we can have shortcuts for those 3 actions.

So, as a fix I would have preferred 2 commits:

* one fixing in-place the existing actions (in order to see the actual meat) and 
* another commit refactoring the classes.

I seems the existing static inner classes might have been sufficient: you could have just made them public. You could have introduced SplitDocumentHorizontallyAction which just extends SplitDocumentAction and calls super(tc, JSplitPane.HORIZONTAL_SPLIT). Same for SplitDocumentVerticallyAction.

I believe initTopComponent only exists so you can have a no-arg constructor which is odd because you are instantiating the actions manually (so it must have to do with the registrations). Also, you know that `(tc instanceof Splitable && ((Splitable)tc).canSplit())` before it's called.

So, what I am curious to learn is how does the shortcut work for your manually created instance?

asfgit commented:
Can one of the admins verify this patch?

Chris2011 commented:
Smth missing here for the merge?

emilianbold commented:
@Chris2011 I don't see any new commits based on my comment from Sep 23rd.

Chris2011 commented:
Most of the code, was only moved out to separated the code and created actions, I only created one base class by my self. You only mean that I should split it up to 2 commits?

The last sentence I don't understand in your comment:

> So, what I am curious to learn is how does the shortcut work for your manually created instance?

emilianbold commented:
I wrote a rather long message in September. Does nothing I said there make sense?

It hard to review a patch that has +320 line and −101 lines in a single commit.

From what I remember and what I deduced while trying to figure it out, you just added an `@actionreference` and some helper class then refactored.

Reviewing would be much simpler if refactoring would be a separate commit.

>The last sentence I don't understand in your comment:
>>So, what I am curious to learn is how does the shortcut work for your manually created instance?

I was just curious how the solution even works considering the action is created manually and not by the Platform. How does the shortcut plumbing work? Because, maybe you don't even need an `@actionreference`?

Chris2011 commented:
> It hard to review a patch that has +320 line and −101 lines in a single commit.
This is what I didn't get yet, I didn't get your point of this. Thx it is clearer.

Yes, I added 3 ActionReferences for 3 new actions "Split Action Horizontally", "Split Action Vertically" and "Clear Split" because those actions didn't exist. Now you can change the hotkeys for each of them, which was not possible before.

I only set such ActionReferences to have Actions, where I can add shortcuts to it. It is from the official Documentation I think, how we can create actions. I don't know an other way. Only the layer.xml but this is obsoleted I think(?)

emilianbold commented:
But if you instantiate the action manually with `new SplitDocumentVerticallyAction` how is the shortcut set? Maybe I'm missing something really obvious.

Chris2011 commented:
The shortcut is set here: https://github.com/apache/incubator-netbeans/pull/1/commits/67bdd60109474c543c010d3f818cfb36abb9c0c9#r150193869

The functionality is still there, if you right click on an editor tab -> Split -> Vertically. So I decoupled the functionality from https://github.com/apache/incubator-netbeans/pull/3/commits/5dc272ebffd95dfd55720ebfdd697d8ff319c862#diff-c2b5538f4bdcf56983c605aebf73887a and created new classes for each action to set the ActionReference.

Now you can use the functionality like before via Right click on editor tab -> Split -> Vertically OR you can use the shortcut, which I set to Ctrl + Alt + Shift + V or you can set your own.

Chris2011 commented:
Please see my latest commits, I splitted my one commit into two as requested. Thx.

Chris2011 commented:
Can someone have a look now, please? :)

dtrebbien commented:
Hello @Chris2011,

I believe what @emilianbold meant by having two commits is that the first commit would implement the requested feature while keeping the `SplitDocumentAction` and `ClearSplitAction` nested classes within `org.netbeans.core.multiview.SplitAction`. Presumably you would need to add `SplitDocumentHorizontallyAction` and `SplitDocumentVerticallyAction` nested classes so that you could add the specific `@ActionID`, `@ActionRegistration`, and `@ActionReference` annotations. Then, once you have done this and verified that you are able to invoke the actions via the keyboard shortcuts, the second commit would be a simple refactoring that would move the `ClearSplitAction`, `SplitDocumentAction`, `SplitDocumentHorizontallyAction`, and `SplitDocumentVerticallyAction` nested classes to top-level classes within the `org.netbeans.core.multiview.actions` package.

One subtle issue is that moving message strings LBL_ClearSplitAction, LBL_ValueClearSplit, LBL_SplitDocumentActionHorizontal, LBL_ValueSplitHorizontal, LBL_SplitDocumentActionVertical, LBL_ValueSplitVertical to a different package (from `org.netbeans.core.multiview` to `org.netbeans.core.multiview.actions`) will break existing translations of these strings. For example, there are Czech translations at https://netbeans.org/projects/nblocalization/downloads/directory/8.0.2 which would no longer work. Maybe these message strings could be kept within the `org.netbeans.core.multiview` package and the relocated `ClearSplitAction`, etc. could refer to `org.netbeans.core.multiview.Bundle` for these strings. Or, maybe it's fine to move the strings to the different package. I am not sure.

sdedic commented on core.multiview/src/org/netbeans/core/multiview/actions/ClearSplitAction.java:
```diff
@@ -0,0 +1,82 @@
+ * @author Christian Lenz (Chrizzly)
+ */
+@ActionID(
+        category = "Tools",
```
Consider category "Windows" or "Windows / Layout" rather than Tools.

sdedic commented on core.multiview/src/org/netbeans/core/multiview/actions/SplitDocumentHorizontallyAction.java:
```diff
@@ -0,0 +1,67 @@
+
+        if (tc != null) {
+            super.setTopComponent(tc);
+            super.setOrientation(JSplitPane.HORIZONTAL_SPLIT);
```
Could be the orientation set from the constructor (it is constant).

sdedic commented on core.multiview/src/org/netbeans/core/multiview/actions/SplitDocumentHorizontallyAction.java:
```diff
@@ -0,0 +1,67 @@
+@ActionRegistration(
+    displayName = "#LBL_ValueSplitHorizontal"
+)
+@ActionReference(path = "Shortcuts", name = "DOS-H")
```
Also consider creating `@ActionReference` for placement in `Menu/Windows`. Note that the registered singletons (unlike instance created in window popup menus) do not update their enable flag according to the current TC.

Chris2011 commented:
@dtrebbien thx for the explanation, the refactoring was needed to added the ActionIds, I tried it before I refactored the code, because that wasn't my intention. So I changed the code a bit and got it working, while I created new actions in external classes. Maybe I made a mistake and it would work right inside as nested classes but now I think it is a bit cleaner.

Chris2011 commented:
@sdedic thx for reviewing, I will have a look later.

emilianbold commented:
BTW, regarding my remark:

> But if you instantiate the action manually with new SplitDocumentVerticallyAction how is the shortcut set? Maybe I'm missing something really obvious.

I believe there may be 2 instances of the same action. One added manually with new which basically does not have a keyboard shortcut and another one added by the NetBeans infra which does have the binding. My confusion was probably assuming there is a single instance.

I would appreciate somebody confirming this or not, etc. because it was not obvious to me.

Chris2011 commented:
Yeah so the actions, which are added manually via `new` are for the contextmenu of an opened tab
![image](https://user-images.githubusercontent.com/795658/35099696-4353e148-fc59-11e7-8832-b2ae01c28dea.png)


and for the menu 

Window -> Configure Window -> Split Document.
![image](https://user-images.githubusercontent.com/795658/35099674-2f5ff712-fc59-11e7-9e5a-56d2fc2eacf1.png)

Those actions where added before now I only create a new instance, but yes maybe it could be done even better. Changes are appreciated too.

Chris2011 commented:
@sdedic markiewb and me had a coding session yesterday. Big thx to him. He helped me a lot with other stuff and with this PR. So in general we fixed everything what was requested (Refactoring, simplifications, etc.). Hope this is now enough to merge the request.

We tested all actions. View -> Split; Window -> Configure Window -> Split Document; Right Click on an editor tab -> Split; And my shortcuts.

@emilianbold Yes those are 2 instances. One for the menu actions and one for my shortcuts via the ActionReference. But both of us, don't see any problem with this. We added some comments too to describe the constructors a bit better.


Cheers

Chris

geertjanw commented:
Great. So @Chris2011, from your and markiewb point of view, everything is now done and there's nothing at all that should block this one from being accepted?

Chris2011 commented:
@geertjanw right. We don't see any other problems with that stuff. Maybe an other can test it before, if someone wants it. But from our pov, everything should be fine.

geertjanw commented:
Would be good for @sdedic and @emilianbold to take a quick look.

emilianbold commented:
Looking at it.

emilianbold commented:
To me this looks like a very clean patch of what this PR is all about: https://github.com/Chris2011/incubator-netbeans/compare/Chris2011:master...emilianbold:feature/248233-doc-split-actions?expand=1 (I've also made a PR against Chris' repository https://github.com/Chris2011/incubator-netbeans/pull/1 ): you just want to register some actions which have some shortcuts and call splitWindow / clearSpit.

Is anything missing if we do it this way?

Any reason SplitAction had to be changed to use the same SplitDocumentHorizontallyAction, etc. other than code 'purity'?

I guess the PR got better with the separate commits but I don't find it good because it changes too much for no reason. Why introduce that tiny abstract class just to store that `int orientation`?  Also having that initTopComponent method and the separate action constructors that accepts a TopComponent seemed odd somehow.

If my changes are OK and work you have my vote for it and are free to merge it.

sdedic commented:
I think it's good; thanks, @Chris2011

emilianbold commented:
PS: Note that SplitAction is registered in the layer and visible under the View menu. Another approach since you create official actions for each would be for that registration to be a normal sub-folder containing your actions... :-)

Chris2011 commented:
The reason for changing was, I changed the code anyway so why not doing a bit of refactoring like decouple them into separate action classes? Sure, the code of NB is huge but why not doing it a bit better than before? The tutorials of creating actions with 3rd-party-plugins, works exactly that way. Create an action class, add the actionreferences, etc.

I thought it could be good to clean up the code a bit, sure maybe it was more code than before but it is more maintainable. It was good to start to refactor the code to bring in consistency. Actions are in action folders (If they make sense). Actions are in separate classes (One class does exactly one thing, SplitHo., SplitVer, ClearSplit, etc.)

I added an Actions folder before but the problem was discussed some comments above, trouble with the translations etc. I've totally forgotten it.

Chris2011 commented:
@emilianbold as I understand it right, you wanted to leave the existing code as it is, right? Only adding those 3 missing actions? When I understand your link correct: https://github.com/Chris2011/incubator-netbeans/compare/Chris2011:master...emilianbold:feature/248233-doc-split-actions?expand=1

If I understand it correct, then we have duplicated code and I wanted to avoid that. If I missunderstand smth, please let me know it.

Chris2011 commented:
But I'm also fine with your code.

Chris2011 commented:
@emilianbold ok I understand your point. Maybe I not finished my thinking about the refactoring part. Anyway, I'm fine with merging your PR into my branch and will push it again. Thx for anything :)

Chris2011 commented:
So now I tested and merged the changes from @emilianbold PR to my PR. Hope now everything is fine.

Chris2011 commented:
So I will wait until next friday and will merge, if it is allowed(?) Please leave a last comment thx :)

geertjanw commented:
Sure, happy to merge this, if @emilianbold gives his final OK, since he's worked with you a lot on this.

sdedic commented:
:+1:

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/AbstractSplitDocumentAction.java:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @author Christian Lenz (Chrizzly)
+ */
+public abstract class AbstractSplitDocumentAction extends AbstractAction {
```
Why does this class still exist?

Chris2011 replied:
Fixed.

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/ClearSplitAction.java:
```diff
@@ -0,0 +1,62 @@
+    "LBL_ClearSplitAction=&Clear",
+    "LBL_ValueClearSplit=clearSplit"
+})
+public final class ClearSplitAction extends AbstractAction {
```
... since this one is a plain abstract action

Chris2011 replied:
Didn't understand, what you mean.

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/SplitAction.java:
```diff
@@ -81,7 +80,7 @@ public JMenuItem getMenuPresenter() {
 	return getSplitMenuItem();
     }
-    
+
```
... these whitespaces don't need to be changed.

Chris2011 replied:
Can't see any whitespace here :/

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/SplitLayerUI.java:
```diff
@@ -93,7 +93,7 @@ public void eventDispatched( AWTEvent event ) {
                         public void run() {
                             TopComponent tc = (TopComponent)SwingUtilities.getAncestorOfClass( TopComponent.class, content );
-                            SplitAction.splitWindow( tc, orientation, splitLocation );
+                            AbstractSplitDocumentAction.splitWindow(tc, orientation, splitLocation);
```
... why not leave the original call, it still works

Chris2011 replied:
Fixed

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/SplitLayerUI.java:
```diff
@@ -40,7 +40,7 @@
 /**
  * Layer that paints a split divider line when the multiview is about to be split using mouse.
- * 
+ *
```
whitespace

Chris2011 replied:
Can't see any whitespace here :/

emilianbold commented on core.multiview/src/org/netbeans/core/multiview/SplitAction.java:
```diff
@@ -112,13 +111,17 @@ static boolean isSplitingEnabled() {
 		if (tc instanceof Splitable && ((Splitable)tc).canSplit()) {
-		    JMenuItem item = new JMenuItem(new SplitDocumentAction(tc, JSplitPane.VERTICAL_SPLIT));
+
+		    JMenuItem item = new JMenuItem(new SplitDocumentVerticallyAction());
```
... not really certain we need to use SplitDocumentVerticallyAction here, the existing SplitDocumentAction works just as well

Chris2011 commented:
Hm, maybe there was smth wrong with the merge :/. Sry about that. Will change and test it later. Thx @emilianbold

Chris2011 commented:
I fixed most of the requested changes now.

Chris2011 commented:
@emilianbold can you check it again please?

emilianbold commented:
Take a look at this diff https://github.com/apache/incubator-netbeans/compare/master...emilianbold:feature/248233-doc-split-actions and compare with this PR's changes https://github.com/apache/incubator-netbeans/pull/3/files

junichi11 commented:
@Chris2011 Maybe, Emi says it is better to add only the necessary changes. e.g. You added some new lines. BTW, you should use `/*` as the first line of license header instead of `/**`.

Chris2011 commented:
I thought I only added that, what emi suggested. So after this, the merge was kind of "broken" so I tried to fix it manually.

Chris2011 commented:
But in my opinion, this shouldn't be so bad to merge it now. It is as it is, but the concept was to make is simple and only do, what needed. And this was done now.

emilianbold commented:
> But in my opinion, this shouldn't be so bad to merge it now. It is as it is, but the concept was to make is simple and only do, what needed. And this was done now.

I personally prefer patches to be tight and not reformat code, or add random lines or whitespace, or fix unrelated bugs, etc. 

The PR is not "so bad" to be un-mergeable right not but you are also 99% done and instead of posting that comment you could have amended the commit and be done with it.

Chris2011 commented:
@emilianbold thx for your opinion and I really know what I mean. I will be more strict in the next PRs and will keep your opinion in mind and everything, what you said. I really can understand what you mean. I will merge the PR now, if there are no other big problems from whoever. Thx for the great discussion.

junichi11 commented:
I'm not sure why you don't change that... 
**NOTE**: You should not merge this using "Merge pull request" button as it is. This PR includes many irrelevant commits.

junichi11 commented:
I suppose you should fix the license header although it is a trivial thing.

@matthiasblaesing Could you help this? I think that Chris should rebase and squash this PR in Chris's local machine or someone should merge it using "Squash and merge" button.

Chris2011 commented:
@junichi11 Sry, I missed the point with the header. Will fix it. Thx. Yeah maybe I need some help with that, what is not working out. I rebased, and my stuff is on top of it.

matthiasblaesing commented:
I tried a classic git rebase - I'll make it short: I aborted it, as it ended with a lot of merge conflicts. The history of this PR is long and winding and this makes it hard.

In this case I would try the github "Squash and merge" variant. This should result in one commit in the history.

You could do it by hand:

- checkout master into a working branch lets call it `pr3-merged`
- make sure you have github PRs fetched (in my case the remote is called github) - if so there should be a branch `github/pr/3` which holds the changes from this PR
- merge `github/pr/3` into `pr3-merged`
- create a diff between `master` and `pr3-merged`: `git diff master..pr3-merged > ../pr3.patch` the output can be outputted into a file and that patch is the core of this PR
- now you could switch your clone back to `master`
- delete the `feature/248233-doc-split-actions` branch locally
- checkout master as a new branch `feature/248233-doc-split-actions`
- apply the patch created above via `git apply ../pr3.patch`
- commit the changes
- force push into the existing `feature/248233-doc-split-actions` in the repository `Chris2011`

If you have done it before, this can be done quickly, if not, I would not start in a big repository like netbeans :-)

junichi11 commented:
@matthiasblaesing Thanks a lot for your help!

> In this case I would try the github "Squash and merge" variant. This should result in one commit in the history.

:+1: 
Could you merge this PR if it's OK with Chris?

BTW, diff of this PR is here: https://github.com/apache/incubator-netbeans/pull/3.diff
So, in this case, Chris might be able to use the following steps:
- Download the diff file as a patch
- Rename the branch: e.g. `git branch -m feature/248233-doc-split-actions-old` 
- Checkout master (see Matthias's comment) 
- Apply the patch (see Matthias's comment)
- Force push (see Matthias's comment. Especially, please note **"in the repository Chris2011"**.)

matthiasblaesing commented:
Chris is a committer, so he can commit himself. @Chris2011 if you need help, feel free to contact me.

junichi11 commented:
Sure. Anyway, I wanted to avoid merging it as it is :)

Chris2011 commented:
Will try that. Thx.

Chris2011 commented:
So now, everyhing should be fine now. I renamed the old branch, made a clean one, updated all branches that is needed, applied the patch, changed the line endings from CRLF to Unix LF. 3 files are new, which are the 3 actions (ClearSplit, SplitDocumentVerticallyAction and SplitDocumentHorizontallyAction). The SplitAction.java file only removes some Bundle messages.

Atm, I don't know why is the SplitAction completely red and green again but finally, it shouldn't matter now. Everything was discussed earlier. Please merge now.

matthiasblaesing commented:
Please recheck SplitAction - it now has windows line endings (\r\n), which is also the reason, that it looks so red.

Chris2011 commented:
This is what I said, I already checked all files, expecially SplitAction. I change the line endings from CRLF to Unix LF before I applied the patch. I checked it doubled after I applied the patch and it was still LF.

Chris2011 commented:
Hm, I closed the file and reopened it, now it is CRLF, will fix it.

Chris2011 commented:
It doesn't make sense to change it now to LF and amend the commit, right? Do I have to do it again? I mean applying the patch and fix the line endings?

matthiasblaesing commented:
If you change line endings, do an amend commit and then force push that should do it.

Chris2011 commented:
Got it.

matthiasblaesing commented:
You now lost the commit message :-) - When you amend you need to keep in mind, that you replace the previous commit.

Chris2011 commented:
Hm, ok. I thoght they will merged together. The Message too. But this should Not be a Problem now. Now I know it and will make it better next time :).

matthiasblaesing commented:
Why wait? An amend commit is possible without any file changes, so you can still fix the commit message. Same procedure s before:  Start the amend commit, check/update the message, finish the commit, force push.

Am 15. September 2018 23:26:34 MESZ schrieb Chrizzly <notifications@github.com>:
>Hm, ok. I thoght they will merged together. The Message too. But this
>should Not be a Problem now. Now I know it and will make it better next
>time :).

Chris2011 commented:
Done.

matthiasblaesing commented:
Looks better, though I would not choose this message. I would have worded it:

```
[NETBEANS-59] Separate actions for document splitting to enable shortcuts

<Context explaining why it was implemented the way it was>
```

It is a summary what was the desired outcome of this commit, which helps to understand the code changes. If an explanation is required why it was implemented the way it was or context should be provided, the further lines of the message can be used. The context is not always required, but helps to put the understand changes years later.

You can change the message or commit as is.
